Generator: Refactor handling of reference tables.

masnax opened the pull request:
## Background
Working on the generator, we noticed there were several table views with the suffix `_ref` (i.e. `instances_devices_ref`) that were used exclusively by the generator. In an effort to make the generator more readable, as well as more extensible, we decided to move handling this logic from the database to `go` itself, thus eliminating a need for the `_ref` table views.

## Objectives
* Remove all usage of `_ref` table views (if all goes well they can be dropped at a later point).
* Use structs that map to table fields to generate exactly one set of methods / statements per supported table.
* Remove context about the project itself from the generator, for example the generator should not know what `Instance` means.

## Implementation
* Added four new structs: `db.Devices`, `db.Config`, `db.InstanceProfiles`, and `db.CertificateProjects` these map to reference tables (not views) for structs/tables currently already supported by the generator. This allows us to explicitly tell the generator what tables we want and what columns are in those tables, as opposed to the current generator logic which is highly coupled to the currently supported tables.
* As these structs' methods need to have different behaviour, I introduced a `TableType` type to the generator which is used to discern context about how to read the struct, and in what form to return the data from the table.
  * `EntityTable` is just a standard, independent entity table. (`instances`, `profiles`, `projects`, etc.)
  * `ReferenceTable` is a struct / table that is referenced by another entity. (`instance_devices`, `profiles_devices`, etc.)
  * `AssociationTable` is a struct / table that associates two other `EntityType`s (`instances_profiles`, `certificates_projects`)
  * `MapTable` is a `ReferenceTable` with key/value entries. (`instances_config`, `instances_devices_config`, etc.)

## Example of changes
* Using devices for instances as an example, previously generated methods for `Instance` would expect the table view `instances_devices_ref` to exist. This view checks several tables (`instances`, `nodes`, `instances_devices`, and `instances_devices_config`). The generated method `GetInstances` queries this view and creates a nested map of the data, which is then added to the `Instance` struct by its `Project` and `ID`. 

* In the new implementation, there are two structs, `Device` and `Config`, which are supported by the generator. These structs would be read in the generator with the `TableType` `ReferenceTable` and `MapTable` respectively, and CRUD methods would be generated for them. These new methods don't check the views, but the actual device and config tables. `GetInstances` would then call `GetDevices` which calls `GetConfig`, (the same is true for `Create/Delete/Update`) and all of the logic in SQL is moved over to go.

## Additions and Changes
* Add `Device` struct to represent all `{entity}_devices` tables to the generator, and generate CRUD methods.
* Add `Config` struct to represent all `{entity}_config` tables to the generator, and generate CRUD methods.
* Add `InstanceProfiles` struct to represent the `instances_profiles`table to the generator, and generate CRUD methods.
* Add `CertificateProjects` struct to represent the `certificates_projects`table to the generator, and generate CRUD methods.
* Remove `{Get/Create/Delete}Ref` generated methods, and instead call methods generated for the above structs.
* Add `TableType` to the generator to support different implementations of CRUD generated methods.

If all goes well, it would be nice to extend this to include an `OrderedListTable` type for `TableType`, to deal with things like `instances_profiles` which has an `apply_order` column.

There's a few hacky things going on here as well, at the moment:
* for `ReferenceTable`s, instead of generating individual statements for each table, there is a generic statement with formatted table names.
* My goal is to eventually harmonize the new `Device` struct with the API-side`deviceConfig.Device`, but for now I'm just converting to the latter wherever possible. This is because while the `Device` struct has a field for `Type`, it's just another config key as far as the API-side is concerned.
* `UpdateInstanceProfile` and `GetProjectUsedBy` are still manual methods until the generator can be extended to support Storage volumes, networks, and `OrderedListTable`.

lxc-jenkins commented:
Testsuite passed

tomponline commented:
>     * `GetInstances` would then call `GetDevices` which calls `GetConfig`, (the same is true for `Create/Delete/Update`) and all of the logic in SQL is moved over to go.

Thanks for updating the PR description, this is one of the best PR descriptions I've ever seen :)
Of course most PRs don't need this level of detail, but I really appreciate it on this one as its covering an area that I am not knowledgable in, so you covering the changes in detail is extremely useful.

WRT the quoted command above, will will this result in lots of smaller queries being sent to the database, or is it converted/built into a single query (like the original view query was)? Also if using lots of queries, are they all done inside a single transaction?

Out of interest, was there any consideration given to just moving the reference table view query into a function that then returned the same result as the view itself, as that would allow us to remove the views themselves (that were making applying DB patches more difficult) with potentially a simpler/smaller approach?

masnax commented:
@tomponline this should be ready for another look-over. I switched over to using `fmt.Errorf` everywhere in the generator, and `Devices` is now a `map[string]Device` instead of `[]Device`. As well, I split up generation of methods like `GetInstanceProfiles` into two methods: `GetInstanceProfiles` which gets profiles for instances, and `GetProfileInstances`, which gets instances for profiles.

tomponline commented:
@masnax this needs a rebase please.

tomponline commented on lxd/db/cluster/constants.go:
```diff
@@ -84,6 +84,14 @@ var EntityFormatURIs = map[int]func(a ...interface{}) string{
+	TypeInstance: func(a ...interface{}) string {
+		uri := fmt.Sprintf(EntityURIs[TypeInstance], a[1], a[0])
+		if a[0] == "default" {
+			return strings.Split(uri, fmt.Sprintf("?project=%s", a[0]))[0]
```
@masnax from what I can see this bit and the URLs generated below haven't been addressed yet (with regards to manually tacking on the URL project parameter).

masnax replied:
> > > Better generation of URIs
> > 
> > Can this be a separate PR by any chance?
>
> Absolutely.

I have the URI changes prepared in a different branch already.

tomponline replied:
Excellent thanks, lets try and remove as much as possible from this PR as its rather large to review properly. Thanks

stgraber replied:
@masnax what's the status on this?

masnax replied:
@stgraber I've removed this type of URL handling everywhere, in favour of the `api.NewURL()` method. I have the branch here, but it needs a small rebase: https://github.com/masnax/lxd/tree/generator/uris

tomponline commented on lxd/db/generate/db/method.go:
```diff
@@ -193,11 +193,7 @@ func (m *Method) getMany(buf *file.Buffer) error {
-	buf.L("}")
-	buf.N()
-
+	m.ifErrNotNil(buf, "nil", fmt.Sprintf("errors.Wrap(err, \"Failed to fetch %s\")", lex.Plural(m.entity)))
```
@masnax looks like `errors.Wrap` is still being used here.

stgraber replied:
@masnax still seeing a bunch of errors.Wrap

masnax replied:
@stgraber I didn't remove them everywhere in the generator, just in `lxd/db/generate/db/method.go`. Since it's an extra commit on this PR anyway, I can just remove `errors.Wrap` everywhere in the generator as one commit.

tomponline commented on lxd/db/profiles.go:
```diff
@@ -47,15 +47,15 @@ type Profile struct {
 	Config      map[string]string
 	Devices     map[string]Device
-	UsedBy      []string
+	Instances   []string `db:"omit=create,update&uri=yes"`
```
@masnax im interested what was your thinking around this change? To the naive reader it seemed like having it called `UsedBy` aligned better.

masnax replied:
`profiles.UsedBy` only holds instance URIs at the moment, while `projects.UsedBy` contains instances, networks, acls, storage volumes, etc. This leaves two options with how to deal with this on the generator side:
* Use `UsedBy []string` with tags like `db:"used-by=instances"` to specify what the slice contains
* Just split apart `UsedBy` into its component entity slices: `Instances', `Networks`, etc.

I went with the latter, although as Networks, NetworkACLs, and Storage Volumes are as yet unsupported by the generator, I left UsedBy in place for projects. I also think it's more appropriate as the database shouldn't have any concept of what `UsedBy` is, and instead we can just fill a `UsedBy` slice when converting to an API struct.

tomponline replied:
I see, I think. TBH I'm not really totally happy with us having multiple ways of generating the `UsedBy` values (as we to today already), some places it uses in-Go logic and multiple queries and others its done at object load time apparently (although I'm struggling to see where the queries are actually done due to this being DB generation logic).

Recently I've been moving away from populating UsedBy at object load time from the DB and instead doing it at render time, as often we don't need to go to the expense of populating UsedBy unless its actually being asked for.

Where can I see the resulting query that this logic implements? Thanks

masnax replied:
So there's 3 parts to this:
* Presence of `profile.Instances` 
  * this makes the generated methods for `Get/Create/UpdateProfile` look for their associated `Get/Create/UpdateProfileInstances` methods. These methods would be defined in `instances_profiles.mapper.go` and be generated by the line `//go:generate mapper method -p db -e instance_profile GetMany struct=Profile` as an example for `GetProfileInstances`.
* `db:omit` tags
  * This prevents `CreateProfile` and `UpdateProfile` from detecting `profile.Instances`, because we don't want the profile to change what its used by.
* `db:uri` tag
  * This tells any of the generated methods that use `profile.Instances` to keep the URI format, and not parse it for just the name.

But to directly answer your question:
* `GetProfiles` in `profiles.mapper.go` calls `GetProfileInstances` in `instance_profiles.mapper.go`, and then calls `GetInstanceURIs` using the `instance_id`s obtained.

tomponline replied:
>     * `GetProfiles` in `profiles.mapper.go` calls `GetProfileInstances` in `instance_profiles.mapper.go`, and then calls `GetInstanceURIs` from `instances.mapper.go` using the `instance_id`s obtained

Great thanks that clears things up for me. So this doesn't sound particularly more efficient that what, say, what we do for ACL's UsedBy:

https://github.com/lxc/lxd/blob/9b84e70a0f0e2bb513f480b7138aaa690dbf97a4/lxd/network_acls.go#L153-L159

I wonder if we should remove the population of UsedBy from these functions and instead have a separately implemented `UsedBy` function for each object type that we can then call to populate the `UsedBy` field when needed. This would be more efficient that loading this always as often its unnecessary.

tomponline commented on lxd/db/projects.go:
```diff
@@ -198,6 +198,45 @@ func (c *ClusterTx) InitProjectWithoutImages(project string) error {
 	return err
 }
 
+// GetProjectUsedBy takes a slice of projects and assigns stuff.
```
"assigns stuff"?

masnax replied:
oops 😛

tomponline commented on lxd/db/storage_volumes.go:
```diff
@@ -72,7 +72,10 @@ WHERE storage_volumes.type = ?
 			Description: r[2].(string),
 			PoolName:    r[3].(string),
 			ProjectName: r[4].(string),
-			NodeID:      r[5].(int64),
```
@masnax can this be in a separate PR? Also I'd like to understand more about why this change is required. Thanks

masnax replied:
To build the URI for storage volumes we need the node id of the storage volume. This field is nullable so either `StorageVolumeArgs.NodeID` needs to be a `*int64` or we need to default `storage_volumes.node_id` to `-1` if it's NULL.

tomponline replied:
Can you handled that via IFNULL in the query?

tomponline replied:
Also, but something for another day, there's no reason that field should be NULLable AFAIK.

masnax replied:
That's certainly possible, but I was trying to get away from using fake `nil` values like `-1`.

tomponline replied:
what value was it before this change?

masnax replied:
Previously two separate queries were being performed: one for volumes with node_ids, and one without.
This is the whole view:
```sql
CREATE VIEW projects_used_by_ref (name,
    value) AS
  SELECT projects.name,
    printf('/1.0/instances/%s?project=%s',
    "instances".name,
    projects.name)
    FROM "instances" JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/images/%s?project=%s',
    images.fingerprint,
    projects.name)
    FROM images JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s&target=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name,
    nodes.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN nodes ON node_id=nodes.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 UNION
  SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 AND storage_volumes.node_id IS NULL UNION
  SELECT projects.name,
    printf('/1.0/profiles/%s?project=%s',
    profiles.name,
    projects.name)
    FROM profiles JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/networks/%s?project=%s',
    networks.name,
    projects.name)
    FROM networks JOIN projects ON project_id=projects.id UNION
  SELECT projects.name,
    printf('/1.0/network-acls/%s?project=%s',
    networks_acls.name,
    projects.name)
    FROM networks_acls JOIN projects ON project_id=projects.id;
```

Here are the relevant parts:
```sql
SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s&target=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name,
    nodes.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN nodes ON node_id=nodes.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 UNION
  SELECT projects.name,
    printf('/1.0/storage-pools/%s/volumes/custom/%s?project=%s',
    storage_pools.name,
    storage_volumes.name,
    projects.name)
    FROM storage_volumes JOIN storage_pools ON storage_pool_id=storage_pools.id JOIN projects ON project_id=projects.id WHERE storage_volumes.type=2 AND storage_volumes.node_id IS NULL UNION
```

tomponline replied:
> That's certainly possible, but I was trying to get away from using fake `nil` values like `-1`

Can you decouple this change from this PR, it would be worth discussing separately. 
When we started going down this route for filtering it made sense, as the inputs were optional.
But with this change it feels like its starting to go too far in using nils to indicate something that isn't optional, and perhaps this is a schema issue that we should either fix or try and mask as early as possible and not propagate it.

And assuming we were going to support nil here, its not been completed yet, as  the query itself is still doing `IFNULL(storage_volumes.node_id, -1)` and further down there's a reference to `v.NodeID < 0`.

tomponline commented on lxd/db/generate/db/method.go:
```diff
@@ -230,7 +230,7 @@ func (m *Method) getMany(buf *file.Buffer) error {
 	buf.L("// Select.")
 	buf.L("err = query.SelectObjects(stmt, dest, args...)")
-	m.ifErrNotNil(buf, "nil", fmt.Sprintf("errors.Wrap(err, \"Failed to fetch %s\")", lex.Plural(m.entity)))
+	m.ifErrNotNil(buf, "nil", fmt.Sprintf("fmt.Errorf(\"Failed to fetch from \\\"%s\\\" table: %%w\", err)", entityTable(m.entity)))
```
You may find using backticks avoids the need for so much escaping.

lxc-jenkins commented:
Testsuite passed

masnax commented on lxd/db/profiles.go:
```diff
@@ -74,17 +59,39 @@ func ProfileToAPI(profile *Profile) *api.Profile {
 	Name    *string
 }
 
+// GetProfileUsedBy returns all the instances that use the given profile.
+func (c *ClusterTx) GetProfileUsedBy(profile Profile) ([]string, error) {
+	usedBy := []string{}
+
+	profileInstances, err := c.GetProfileInstances()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instanceID := range profileInstances[profile.ID] {
+		instanceURIs, err := c.GetInstanceURIs(InstanceFilter{ID: &instanceID})
+		if err != nil {
+			return nil, err
+		}
+
+		usedBy = append(usedBy, instanceURIs...)
+	}
+
+	return usedBy, nil
+}
+
```
@tomponline Alright, I've switched `Profiles.Instances` back into `Profiles.UsedBy` and added this manual method `GetProfileUsedBy`.

For now I've left `UsedBy` in place for `Project` and `Profile`, but I've made the `GetProfileUsedBy` and `GetProjectUsedBy` methods just return what one object is used by, so that it's easy to remove the field and get `UsedBy`s only when we want them in a different PR.

tomponline replied:
Thanks will look at this shortly.

tomponline commented on lxd/db/network_acls.go:
```diff
@@ -312,3 +314,24 @@ func (c *Cluster) DeleteNetworkACL(id int64) error {
+	}
+
+	code := cluster.EntityTypes["network acl"]
+	formatter := cluster.EntityFormatURIs[code]
```
@masnax not for this PR probably, but I can see us removing ` cluster.EntityFormatURIs` and `cluster.EntityURIs` in the near future, as URL generation really doesn't need to be in the DB layer.

tomponline commented on lxd/db/storage_volumes.go:
```diff
@@ -1056,3 +1058,46 @@ WHERE storage_volumes.type = ? AND projects.name = ?
+		uri := fmt.Sprintf(cluster.EntityURIs[code], info.PoolName, "custom", info.Name, project)
+		formChar := "&"
+		if project == "default" {
+			uri = strings.Split(uri, fmt.Sprintf("?project=%s", project))[0]
```
@masnax please can you use the api.URL builder for this part

https://github.com/lxc/lxd/blob/master/shared/api/url.go

e.g. 

https://github.com/lxc/lxd/blob/6e377a8121ebe41f668183191c3bc9787bd35df1/lxd/network/driver_common.go#L1046

masnax replied:
So I have this branch that fixes up all the URI handling in the generator: https://github.com/masnax/lxd/tree/generator/uris

I haven't put it up as PR yet so as to not pollute the list of PRs, but I can do so if that would be better.

stgraber replied:
Ah, okay, considering that as response for an earlier question too.

tomponline commented on lxd/db/instances_test.go:
```diff
@@ -202,7 +216,13 @@ func TestInstanceList(t *testing.T) {
 	err = cluster.InstanceList(nil, func(dbInst db.Instance, p db.Project, profiles []api.Profile) error {
 		dbInst.Config = db.ExpandInstanceConfig(dbInst.Config, profiles)
-		dbInst.Devices = db.ExpandInstanceDevices(deviceConfig.NewDevices(dbInst.Devices), profiles).CloneNative()
+		for _, p := range profiles {
```
@masnax why isnt this using `db.ExpandInstanceDevices` anymore?

stgraber replied:
@masnax

stgraber replied:
Answered in a latter comment.

tomponline commented on lxd/network/network_utils_sriov.go:
```diff
@@ -70,17 +69,25 @@ func SRIOVGetHostDevicesInUse(s *state.State) (map[string]struct{}, error) {
 		// Expand configs so we take into account profile devices.
 		dbInst.Config = db.ExpandInstanceConfig(dbInst.Config, profiles)
-		dbInst.Devices = db.ExpandInstanceDevices(deviceConfig.NewDevices(dbInst.Devices), profiles).CloneNative()
+		for _, p := range profiles {
```
@masnax why isnt this using `db.ExpandInstanceDevices` anymore?

I'm also confused why for the same set of inputs the new function can now return an error whereas before it could not.

masnax replied:
There's currently a lot of different types representing a set of Devices: 
* `deviceConfig.Device` which is a mask on `map[string]map[string]string`.
* `map[string]db.Device` which is the new database type for devices.
* `map[string]map[string]string` which is the old database type for devices.

I want to consolidate all these, but it would make this PR too big. For now,`db.ExpandInstanceDevices` returns a `deviceConfig.Device`, and modifying its arguments and return type would again increase the size of the PR. Just keeping the method in use everywhere would introduce a lot of redundant back and forth converting between the new and old types, so instead I'm temporarily manually expanding devices in some places. The goal is to eventually revert back to `db.ExpandInstanceDevices`, but with slightly different arguments and return types. 

As for the new error, previously `deviceConfig.NewDevices` was used to convert from the database representation of Devices: `map[string]map[string]string` to `deviceConfig.Devices`, but there was no actual validation to see if the received map is actually a valid device. Now I've introduced the `db.Device` struct which contains explicitly named fields with defined types like`db.DeviceType`, so validation can be performed.

stgraber replied:
@masnax can we maybe have a FIXME or TODO so it's clear that it's a temporary step and more work is coming?

masnax replied:
@stgraber sure I'll add TODOs for those.

tomponline commented on lxd/project/permissions.go:
```diff
@@ -1043,20 +1067,27 @@ func expandInstancesConfigAndDevices(instances []db.Instance, profiles []db.Prof
 		}
 
+		expandedDevices := db.ExpandInstanceDevices(
+			deviceconfig.NewDevices(db.DevicesToAPI(instance.Devices)), apiProfiles)
```
No need for a newline here.

tomponline commented on lxd/db/instances.go:
```diff
@@ -93,17 +73,18 @@ type Instance struct {
 	Name    *string
-	Node    *string
-	Type    *instancetype.Type `db:"omit=profiles-ref,config-ref,devices-ref"`
+	Node    *string `db:"omit=URIs"` // TODO: Implement multiple joins for URI generation.
```
Is this still TODO now we are moving the UsedBy (and hence the need for URL generation) out of the DB generator?

masnax replied:
This TODO is addressed in my URI branch, so as far as this PR is concerned it's still a TODO.

tomponline reviewed (commented):
Is it possible to break out the Device type and helper functions into a separate PR?

I'm really keen to try and reduce the size of the PR as it feels like its changing too much right now.

tomponline commented:
@masnax also this PR needs a rebase please.

masnax commented:
> Is it possible to break out the Device type and helper functions into a separate PR?
> 
> I'm really keen to try and reduce the size of the PR as it feels like its changing too much right now.

The introduction of the `Config` and `Device` structs are intrinsic to the changes in this PR. Previously, information representing both the database and go sides of fields like `entity.Devices` and `entity.Config` were hard-coded into the generator. The introduction of these structs is what lets us tell the generator what we want from the database, and how to represent it in go.

One thing I could do is split apart the changes on the generator side from the changes on the LXD side. However, this would mean the generator would be broken in the first PR, making it so running `make update-schema` would overwrite the refactor.

tomponline commented:
@masnax so has everything been removed from this PR with regards to the DB generator populating the UsedBy fields now?

masnax commented:
> @masnax so has everything been removed from this PR with regards to the DB generator populating the UsedBy fields now as we discussed?

Right, the generated methods for UsedBy fields are now all gone. `Get[Entity]` methods will still expect that a `Get[Entity]UsedBy` method exists, but it must be manually written.

This will make it easy to later remove UsedBy fields from the database structs entirely.

tomponline commented:
> > @masnax so has everything been removed from this PR with regards to the DB generator populating the UsedBy fields now as we discussed?
> 
> Right, the generated methods for UsedBy fields are now all gone. `Get[Entity]` methods will still expect that a `Get[Entity]UsedBy` method exists, but it must be manually written.
> 
> This will make it easy to later remove UsedBy fields from the database structs entirely.

OK good thats important.

There's a code conflict to resolve im afraid.

masnax commented:
> > > @masnax so has everything been removed from this PR with regards to the DB generator populating the UsedBy fields now as we discussed?
> > 
> > 
> > Right, the generated methods for UsedBy fields are now all gone. `Get[Entity]` methods will still expect that a `Get[Entity]UsedBy` method exists, but it must be manually written.
> > This will make it easy to later remove UsedBy fields from the database structs entirely.
> 
> OK good thats important.
> 
> There's a code conflict to resolve im afraid.

Should be taken care of now.

tomponline reviewed (approved):
I still can't say I totally understand everything that this PR is doing or the wider implications of the changes, but I've spent a long time reviewing it now and @masnax has said it cannot be broken up into smaller more digestable PRs.

The root problem for me is that I don't understand the DB generator's parlance (especially commits like this https://github.com/lxc/lxd/pull/9364/commits/3542a9a29e19bf8eda7b4d3a0c71c9532c9c808d), and so many of the concepts are foreign to me.

@stgraber please can you take a look through this and check you're happy also, as it affects core DB functionality.

lxc-jenkins commented:
Testsuite passed

stgraber commented:
jenkins: test this please

stgraber commented on lxd/db/instance_profiles.go:
```diff
@@ -30,3 +34,43 @@ type InstanceProfileFilter struct {
+	project := instance.Project
+	enabled, err := projectHasProfiles(c.tx, project)
+	if err != nil {
+		return errors.Wrap(err, "Check if project has profiles")
```
Is it one of those cases where we can't use fm.Errorf with %w? If not, we should really stay away from pkg/errors.

masnax replied:
@stgraber, this is just a remnant of copy/pasting. I'll switch it.

stgraber commented:
Looking good to me, looks like most comments were addressed and most of the open ones have fixes ready for submission after this initial PR (thinking about the URI and some of the device stuff at least).

stgraber commented:
@tomponline do you see any blocker? If not, I'd have @masnax sort out that remaining comment and then we can merge this and start reviewing the URI branch.

tomponline commented:
> @tomponline do you see any blocker? If not, I'd have @masnax sort out that remaining comment and then we can merge this and start reviewing the URI branch.

No blockers from me. Thanks for taking a look.

stgraber commented:
jenkins: test this please

stgraber commented:
@masnax 

```
lxd/db/generate/file/write.go:49:13: result of fmt.Errorf call not used
```

stgraber commented:
```
lxd/device/disk.go:254:22: error strings should not be capitalized or end with punctuation or a newline
```

stgraber commented:
Oh, not your fault, that came from @markylaing :)

stgraber commented:
will retest this branch once https://github.com/lxc/lxd/pull/9615 is merged

stgraber commented:
@masnax can you rebase on master, that should take care of this failure.
